File: bellmanFord.cpp
#include <iostream>
#include <vector>
using namespace std;
const int N = 1000+21;
const int INF = 1000000000+21;
int n, m, a, b, c;
int dist[N];
vector<pair<pair<int, int>, int> > edges;

bool bellmanFord(int v)
{
    dist[v] = 0;

    // update the distances
    for (int i=1; i<=n-1; i++)
    {
        for (int j=0; j<m; j++)
        {
            int u, v, weight;
            u = edges[j].first.first;
            v = edges[j].first.second;
            weight = edges[j].second;

            if (dist[u] != INF && dist[u] + weight < dist[v])
            {
                dist[v] = dist[u] + weight;
            }
        }
    }

    // check for negative cycles
    for (int j=0; j<m; j++)
    {
        int u, v, weight;
        u = edges[j].first.first;
        v = edges[j].first.second;
        weight = edges[j].second;

        if (dist[u] != INF && dist[u] + weight < dist[v])
        {
            return false;
        }
    }
    return true;
}

int main()
{
    cin>>n>>m;

    for (int i=1; i<=n; i++)
        dist[i] = INF;

    for (int i=0; i<m; i++)
    {
        cin>>a>>b>>c;
        edges.push_back({{a, b}, c});
    }

    bool ans = bellmanFord(1);

    if (ans) 
    {
        for (int i=1; i<=n; i++)
        {
            if (dist[i] == INF)
                cout<<"INF ";
            else
                cout<<dist[i]<<" ";
        }
        cout<<endl;
    }
    else
    {
    	cout<<"Contains negative cycle\n";
    }
}

File: implementation/bellman_ford.py
INF = int(1e9) + 21
N = 1000 + 21

n, m = map(int, input().split())
dist = [INF] * (N)
edges = []

for _ in range(m):
    a, b, c = map(int, input().split())
    edges.append(((a, b), c))

def bellman_ford(start):
    dist[start] = 0

    for _ in range(n - 1):
        for (u, v), weight in edges:
            if dist[u] != INF and dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight

    # Check for negative-weight cycles
    for (u, v), weight in edges:
        if dist[u] != INF and dist[u] + weight < dist[v]:
            return False
    return True

if bellman_ford(1):
    for i in range(1, n + 1):
        if dist[i] == INF:
            print("INF", end=" ")
        else:
            print(dist[i], end=" ")
    print()
else:
    print("Contains negative cycle")
